server/service/src/sync/translation_and_integration.rs

```diff
@@ -0,0 +1,206 @@
+            connection,
+            sync_buffer,
+        }
+    }
+
+    fn translate_sync_record(
+        &self,
+        sync_record: &SyncBufferRow,
+        translators: &SyncTanslators,
+    ) -> Result<Option<IntegrationRecords>, anyhow::Error> {
+        let mut translators_iter = translators.iter();
+
+        let result = loop {
+            let translator = match translators_iter.next() {
+                Some(translator) => translator,
+                None => break None,
+            };
+
+            if let Some(records) = translator.try_translate_pull(self.connection, &sync_record)? {
+                break Some(records);
+            }
+        };
+
+        Ok(result)
+    }
+
+    pub(crate) fn translate_and_integrate_sync_records(
+        &self,
+        sync_records: Vec<SyncBufferRow>,
+    ) -> Result<TranslationAndIntegrationResults, RepositoryError> {
+        let mut result = TranslationAndIntegrationResults::new();
+        let translators = all_translators();
+
+        for sync_record in sync_records {
+            // Try translate
+            let translation_result = match self.translate_sync_record(&sync_record, &translators) {
+                Ok(translation_result) => translation_result,
+                // Record error in sync buffer and in result, continue to next sync_record
+                Err(translation_error) => {
+                    self.sync_buffer
+                        .record_integration_error(&sync_record, &translation_error)?;
+                    result.insert_error(
+                        &sync_record.table_name,
+                        Some(&sync_record),
+                        None,
+                        translation_error,
+                    );
+                    // Next sync_record
+                    continue;
+                }
+            };
+
+            let integration_records = match translation_result {
+                Some(integration_records) => integration_records,
+                // Record translator not found error in sync buffer and in result, continue to next sync_record
+                None => {
+                    let error = anyhow::anyhow!("Translator for record not found");
+                    self.sync_buffer
+                        .record_integration_error(&sync_record, &error)?;
+                    result.insert_error(&sync_record.table_name, Some(&sync_record), None, error);
+                    // Next sync_record
+                    continue;
+                }
+            };
+
+            // Integrate
+            let integration_result = integration_records.integrate(self.connection);
+            match integration_result {
+                Ok(_) => {
+                    self.sync_buffer
+                        .record_successfull_integration(&sync_record)?;
+                    result.insert_success(&sync_record.table_name)
+                }
+                // Record database_error in sync buffer and in resul
```

Comment:
```suggestion
                // Record database_error in sync buffer and in result
```

Reply:
Thanks: fb81c4caf64d20a49a30bd5d59e1499d71226b7c  in https://github.com/openmsupply/open-msupply/issues/338